File: StackQueue.js
class Note {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class Stack {
  constructor(value) {
    if (!value) {
      this.top = null;
      this.height = 0;
      return;
    }

    const newNode = new Note(value);
    this.top = newNode;
    this.height = 1;
  }

  printStack() {
    if (this.top === null) {
      console.log('Empty Stack');
      return;
    }

    let temp = this.top;
    while (temp !== null) {
      console.log(temp.value);
      temp = temp.next;
    }
  }

  getTop() {
    if (this.top === null) {
      console.log('Top: null');
      return;
    }
    console.log('Top: ' + this.top.value);
  }

  getHeight() {
    console.log('Height: ' + this.height);
  }

  push(value) {
    const newNode = new Note(value);

    newNode.next = this.top;
    this.top = newNode;
    this.height++;
  }

  pop() {
    if (!this.top) return undefined;

    let temp = this.top;
    this.top = this.top.next;
    temp.next = null;
    this.height--;
    return temp;
  }

  get(index) {
    if (index < 0 || index >= this.height) return undefined;

    let temp = this.top;
    for (let i = 0; i < index; i++) {
      temp = temp.next;
    }
    return temp;
  }
}

class Queue {
  constructor(value) {
    if (!value) {
      this.first = null;
      this.last = null;
      this.length = 0;
      return;
    }

    const newNode = new Note(value);
    this.first = newNode;
    this.last = newNode;
    this.length = 1;
  }

  printQueue() {
    if (this.first === null) {
      console.log('Empty Queue');
      return;
    }

    let temp = this.first;
    while (temp !== null) {
      console.log(temp.value);
      temp = temp.next;
    }
  }

  getFirst() {
    if (this.first === null) {
      console.log('First: null');
      return;
    }
    console.log('First: ' + this.first.value);
  }

  getLast() {
    if (this.last === null) {
      console.log('Last: null');
      return;
    }
    console.log('Last: ' + this.last.value);
  }

  getLength() {
    console.log('Length: ' + this.length);
  }

  enqueue(value) {
    const newNode = new Note(value);

    if (!this.first) {
      this.first = newNode;
      this.last = newNode;
      this.length++;
      return;
    }

    this.last.next = newNode;
    this.last = newNode;
    this.length++;
  }

  dequeue() {
    if (!this.first) return undefined;

    let temp = this.first;
    if (this.length === 1) {
      this.first = null;
      this.last = null;
    } else {
      this.first = this.first.next;
      temp.next = null;
    }

    this.length--;
    return temp;
  }

  get(index) {
    if (index < 0 || index >= this.length) return undefined;

    let temp = this.first;
    for (let i = 0; i < index; i++) {
      temp = temp.next;
    }
    return temp;
  }
}

console.log('Stack:');

let myStack = new Stack(1);
myStack.push(2);
myStack.push(3);
myStack.push(4);
myStack.getTop();
myStack.getHeight();
myStack.printStack();

console.log('Queue:');

let myQueue = new Queue(1);
myQueue.enqueue(2);
myQueue.enqueue(3);
myQueue.enqueue(4);
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();
myQueue.getFirst();
myQueue.getLast();
myQueue.getLength();
myQueue.printQueue();

File: StackQueue.cpp
#include <iostream>
using namespace std;

class Node {
  public:
    int value;
    Node* next;

    Node(int val) {
        value = val;
        next = nullptr;
    }
};

class Stack {
  private:
    Node* top;
    int height;

  public:
    /* Constructor */
    Stack() {
        top = nullptr;
        height = 0;
    }

    Stack(int value) {
        Node* newNode = new Node(value);
        top = newNode;
        height = 1;
    }

    /* Destructor */
    ~Stack() {
        Node* temp = top;
        while (temp != nullptr) {
            top = top->next;
            delete temp;
            temp = top;
        }
    }

    /* Methods */
    void printStack() {
        if (top == nullptr) {
            cout << "Empty Stack" << endl;
            return;
        }

        Node* currentNode = top;
        while (currentNode != nullptr) {
            cout << currentNode->value << " ";
            currentNode = currentNode->next;
        }
        cout << endl;
    }

    void getTop() {
        if (top != nullptr) {
            cout << "Top: " << top->value << endl;
            return;
        }
        cout << "Top: nullptr." << endl;
    }

    void getHeight() {
        cout << "Height: " << height << endl;
    }

    void push(int value) {
        Node* newNode = new Node(value);
        newNode->next = top;
        top = newNode;
        height++;
    }

    int pop() {
        if (top == nullptr) return INT16_MIN;

        Node* temp = top;
        int poppedValue = top->value;
        top = top->next;
        temp->next = nullptr;
        height--;
        delete temp;

        return poppedValue;
    }

    Node* get(int index) {
        if (index < 0 || index >= height) return nullptr;
        Node* temp = top;

        for (int i = 0; i < index; i++) {
            temp = temp->next;
        }

        return temp;
    }
};

class Queue {
  private:
    Node* first;
    Node* last;
    int length;

  public:
    /* Constructor */
    Queue() {
        first = nullptr;
        last = nullptr;
        length = 0;
    }

    Queue(int value) {
        Node* newNode = new Node(value);
        first = newNode;
        last = newNode;
        length = 1;
    }

    /* Destructor */
    ~Queue() {
        Node* temp = first;
        while (temp != nullptr) {
            first = first->next;
            delete temp;
            temp = first;
        }
    }

    /* Methods */
    void printQueue() {
        if (first == nullptr) {
            cout << "Empty Queue" << endl;
            return;
        }

        Node* currentNode = first;
        while (currentNode != nullptr) {
            cout << currentNode->value << " ";
            currentNode = currentNode->next;
        }
        cout << endl;
    }

    void getFirst() {
        if (first != nullptr) {
            cout << "First: " << first->value << endl;
            return;
        }
        cout << "Top: nullptr." << endl;
    }

    void getLast() {
        if (last != nullptr) {
            cout << "Last: " << last->value << endl;
            return;
        }
        cout << "last: nullptr." << endl;
    }

    void getLength() {
        cout << "Length: " << length << endl;
    }

    void enqueue(int value) {
        Node* newNode = new Node(value);
        if (first == nullptr) {
            first = newNode;
            last = newNode;
            length++;
            return;
        }

        last->next = newNode;
        last = newNode;
        length++;
    }

    Node* dequeue() {
        if (first == nullptr) return nullptr;

        Node* temp = first;
        if (length == 1) {
            first = nullptr;
            last = nullptr;

        } else {
            first = first->next;
            temp->next = nullptr;
        }

        length--;
        return temp;
    }

    Node* get(int index) {
        if (index < 0 || index >= length) return nullptr;
        Node* temp = first;

        for (int i = 0; i < index; i++) {
            temp = temp->next;
        }

        return temp;
    }
};

int main() {
    Stack* myStack = new Stack();
    myStack->push(1);
    myStack->push(2);
    myStack->push(3);
    myStack->getTop();
    myStack->getHeight();
    myStack->printStack();

    return 0;
}